module/zstd_zfs/zstd.c

```diff
@@ -0,0 +1,562 @@
+		*level = zstdlevel;
+	}
+
+	return (0);
+}
+
+int
+zstd_decompress_level(void *s_start, void *d_start, size_t s_len, size_t d_len,
+    uint8_t *level)
+{
+	const char *src = s_start;
+	uint32_t bufsiz = BE_IN32(src);
+	uint32_t levelcookie = BE_IN32(&src[sizeof (bufsiz)]);
+	uint8_t zstdlevel = zstd_cookie_to_enum(levelcookie);
+
+	ASSERT3U(d_len, >=, s_len);
+	ASSERT3U(zstdlevel, !=, ZIO_ZSTDLVL_INHERIT);
+
+	/* invalid compressed buffer size encoded at start */
+	if (bufsiz + sizeof (bufsiz) > s_len) {
+		return (1);
+	}
+
+	/*
+	 * Returns 0 on success (decompression function returned non-negative)
+	 * and non-zero on failure (decompression function returned negative.
+	 */
+	if (ZSTD_isError(real_zstd_decompress(
+	    &src[sizeof (bufsiz) + sizeof (levelcookie)], d_start, bufsiz,
+	    d_len))) {
+		return (1);
+	}
+
+	if (level != NULL) {
+		*level = zstdlevel;
+	}
+
+	return (0);
+}
+
+int
+zstd_decompress(void *s_start, void *d_start, size_t s_len, size_t d_len, int n)
+{
+
+	return (zstd_decompress_level(s_start, d_start, s_len, d_len, NULL));
+}
+
+static size_t
+real_zstd_compress(const char *source, char *dest, int isize, int osize,
+    int level)
+{
+	size_t result;
+	ZSTD_CCtx *cctx;
+
+	ASSERT3U(level, !=, 0);
+	if (level == ZIO_COMPLEVEL_DEFAULT) {
+		level = ZIO_ZSTD_LEVEL_DEFAULT;
+	}
+	if (level == ZIO_ZSTDLVL_DEFAULT) {
+		level = ZIO_ZSTD_LEVEL_DEFAULT;
+	}
+
+	cctx = ZSTD_createCCtx_advanced(zstd_malloc);
+	/*
+	 * out of kernel memory, gently fall through - this will disable
+	 * compression in zio_compress_data
+	 */
+	if (cctx == NULL) {
+		return (0);
+	}
+
+	result = ZSTD_compressCCtx(cctx, dest, osize, source, isize, level);
+
+	ZSTD_freeCCtx(cctx);
+	return (result);
+}
+
+
+static size_t
+real_zstd_decompress(const char *source, char *dest, int isize, int maxosize)
+{
+	size_t result;
+	boolean_t emerg = B_FALSE;
+	ZSTD_DCtx *dctx;
+
+	dctx = ZSTD_createDCtx_advanced(zstd_malloc);
+	if (dctx == NULL) {
+		if (zstd_dctx_emerg.ptr == NULL) {
+			return (ZSTD_error_memory_allocation);
+		}
+		emerg = B_TRUE;
+		mutex_enter(&zstd_dctx_emerg.mtx);
+		ZSTD_DCtx_reset(zstd_dctx_emerg.ptr,
+		    ZSTD_reset_session_and_parameters);
+		dctx = (ZSTD_DCtx *)&zstd_dctx_emerg.ptr;
+	}
+
+	result = ZSTD_decompressDCtx(dctx, dest, maxosize, source, isize);
+
+	if (emerg == B_TRUE) {
+		mutex_exit(&zstd_dctx_emerg.mtx);
+	} else {
+		ZSTD_freeDCtx(dctx);
+	}
+
+	return (result);
+}
+
+extern void *
+zstd_alloc(void *opaque __unused, size_t size)
+{
+	size_t nbytes = sizeof (struct zstd_kmem) + size;
+	struct zstd_kmem *z = NULL;
+	enum zstd_kmem_type type;
+	int i;
+
+	type = ZSTD_KMEM_UNKNOWN;
+	for (i = 0; i < ZSTD_KMEM_COUNT; i++) {
+		if (nbytes <= zstd_cache_size[i].kmem_size) {
+			type = zstd_cache_size[i].kmem_type;
+			z = kmem_cache_alloc(zstd_kmem_cache[type],
+			    KM_NOSLEEP);
+			break;
+		}
+	}
+	/* No matching cache */
+	if (type == ZSTD_KMEM_UNKNOWN) {
+		z = kmem_alloc(nbytes, KM_NOSLEEP);
```

Comment:
This code was written for BSD and Solaris that do not have that restriction. Adapting kmem_alloc to use your vmem_alloc or whatever in the SPL is likely the right thing to do here, rather than add linux-only code to the ZFS code.

Reply:
@allanjude there is some related work going on here #9034

Reply:
+1 in this case
portability is more universal in future needs instead of lock to one architecture.

Reply:
right now this part of the code is not portable since kmem_alloc has a different behaviour on Linux/SPL than on BSD/Solaris. Thats a deadlock and must be implemented architecture specific.  btw. the code fails with invalid zstd compression level while running the tests. the level doesnt seem to be preinitialized correct